### PR TITLE
Director-based origin health tests

### LINKED ISSUE #175: Director-based health tests
In #167, we add some simple origin-based self-tests.  However, this only really touches the local host -- there's no indication whether the origin is open to the outside world.

The director has a list of all active origins and there's a federation key pair we can use to generate transfer tokens.  Let's create a new set of tests that have the director periodically upload/download from all the connected origins.

The results can be polled from the origin itself (allowing them to be incorporated into the origin's local monitoring dashboard) and exposed via Prometheus.

### PR DESCRIPTION
**Pull request by haoming29**
Closes #175 

## Test Instructions

Make sure in your `pelican.yaml`, the host of your `Federation.DirectorUrl` is the same as your `hostname`. If you run things in container, it should be a string instead of `localhost`

Have director & registry & origin running at debug mode (`-d`) and watch for the log message at director's end: 

```console
DEBUG[2023-11-07T14:48:11Z] Starting Director test for origin fae8c2865de4 at https://fae8c2865de4:8443 
DEBUG[2023-11-07T14:48:41Z] Starting a new Director test cycle for origin: fae8c2865de4 at https://fae8c2865de4:8443 
......
DEBUG[2023-11-07T14:48:41Z] Director test cycle succeeded at Tue Nov  7 14:48:41 UTC 2023
```

Also, at origin's end, the console message should reflect that the director uploaded/downloaded/deleted a file to/from origin and also you **shouldn't** see any errors from Xrootd suggesting a file upload is failed:

```console
INFO[2023-11-07T14:48:41Z] Served Request                                client=172.17.0.2 daemon=gin fields.time=1.792959ms method=POST resource=/api/v1.0/origin-api/directorTest status=200
```

### CODE REVIEW

**Comment by haoming29** [comment]
When implementing this, I found that for director's scitoken to be verified by XRootD, its public key needs to be part of the jwks generated at the start of the xrootd (ref: https://github.com/PelicanPlatform/pelican/blob/42b92154b39a5304d84b34fc5f7ab40df1e7b039/xrootd/xrootd_config.go#L185) but it's impossible right now for us to manipulate this jwks as it's a pointer to interface. I will need #295 to be merged first which addressed this issue.

**Review by turetske — CHANGES_REQUESTED** [review]
I have a few changes to request. I definitely think someone else should look over this as well, as it's a large PR and I'm sure I might have missed some things.

File: cmd/director_serve.go
Comment (haoming29) [reply cmd/director_serve.go:64]: This should be a tragedy after a very large a merge conflict. Thanks for catching it!

File: origin_ui/origin_api.go
Comment (haoming29) [reply origin_ui/origin_api.go:58]: Yeah... This handler for right now only serves as a check for director's health test report, so I don't think we need to check authz. The caller check is at here around line 75 where `valid, err := director.VerifyDirectorTestReportToken(token)` checks if the token `issuer` is the director. With that being said, I think I will rename this handler to `directorApiAuthHandler` or something like that and ensure it's only used for checking requests from the director.

File: director/origin_api.go
Comment (haoming29) [reply director/origin_api.go:196]: That's good to know. I will add WLCG components to the token. If this is the future, then token generated in #289 should also adhere to the WLCG profile.

Comment (haoming29) [reply director/origin_api.go:196]: Looking at the link you pointed to me, I thought this is the token mainly for file transfer and we had a bunch of scope claims that are only for internal API calls, i.e. advertise/monitoring/reporting like `Claim("scope", "pelican.advertise")`. How can we adhere to WLCG profile while having our unique scopes?

Comment (turetske) [reply director/origin_api.go:196]: Good point. Let's leave it for now.

**Comment by haoming29** [comment]
I still can't make origin happy when accepting director's test files, with Brian's patch merged. I added director as an issuer to scitoken config. I tried to add director's public JWK but it turned out to be the same as the origin's JWK (which makes sense as they are generated based off of the same private key). Adding director's public JWK or not generates the same error as following:

**Director:**
```console
DEBUG[2023-11-06T17:26:34Z] Starting Director test for origin fae8c2865de4 at https://fae8c2865de4:8443 
DEBUG[2023-11-06T17:26:34Z] Starting a new Director test cycle for origin: fae8c2865de4 at https://fae8c2865de4:8443 
WARNING[2023-11-06T17:26:34Z] Director test cycle failed during test upload: Error response 403 from director test upload: 403 Forbidden 
```

**Director's generated scitoken:**
```
eyJhbGciOiJFUzI1NiIsImtpZCI6IjV3NE1ZYTRPUnFndjk4eUZleFpDZXFPZklacTE0QlRBMUJqX2w3Z0hDRVEiLCJ0eXAiOiJKV1QifQ.eyJhdWQiOlsiaHR0cHM6Ly9mYWU4YzI4NjVkZTQ6ODQ0MyJdLCJleHAiOjE2OTkyOTE2NTQsImlhdCI6MTY5OTI5MTU5NCwiaXNzIjoiaHR0cHM6Ly9sb2NhbGhvc3Q6ODg4OCIsImp0aSI6IlBuX0l2STB6RVZaNXNvVVFwaEVibWciLCJzY29wZSI6InN0b3JhZ2UucmVhZDovIHN0b3JhZ2UubW9kaWZ5Oi8iLCJzdWIiOiJkaXJlY3RvciIsIndsY2cudmVyIjoiMS4wIn0.NwaPezTY6NCi2oAuS8DJ_IKG_7_vXN8XuLO3KpveyUKgyIAVJP_0GwONv9Sa0so2hp8hkYsg85ABWIbW34zAEw
```

`/run/pelican/xrootd/export/.well-known/issuer.jwks`
```json
{
 "keys": [
  {
   "alg": "ES256",
   "crv": "P-256",
   "kid": "5w4MYa4ORqgv98yFexZCeqOfIZq14BTA1Bj_l7gHCEQ",
   "kty": "EC",
   "x": "LW_iiCcUz0kBnqQnVWQDlq4wsZu4A0u6ClH9kSfxOwU",
   "y": "KqXC0Lfx3BMsuPJWznQD00Q8xmafv56XeBrqF7DjyOk"
  }
 ]
}
```

And I confirm that the signature of the token can be verified by the public key

**Origin**
```console
INFO[2023-11-06T17:26:34Z] 231106 17:26:34 3218 scitokens_GenerateAcls: Failed to deserialize SciToken: SSL peer certificate or SSH remote key was not OK  daemon=xrootd
INFO[2023-11-06T17:26:34Z] 231106 17:26:34 3218 scitokens_Access: Failed to generate ACLs for token  daemon=xrootd
INFO[2023-11-06T17:26:34Z] 231106 17:26:34 3218 acc_Audit: unknown.2:36@fae8c2865de4 deny https *@[::ffff:172.17.0.2] create /pelican/monitoring/director-test-2023-11-06T17:26:34Z.txt  daemon=xrootd
INFO[2023-11-06T17:26:34Z] 231106 17:26:34 3218 scitokens_GenerateAcls: Failed to deserialize SciToken: SSL peer certificate or SSH remote key was not OK  daemon=xrootd
INFO[2023-11-06T17:26:34Z] 231106 17:26:34 3218 scitokens_Access: Failed to generate ACLs for token  daemon=xrootd
INFO[2023-11-06T17:26:34Z] 231106 17:26:34 3218 acc_Audit: unknown.2:36@fae8c2865de4 deny https *@[::ffff:172.17.0.2] excl_create /pelican/monitoring/director-test-2023-11-06T17:26:34Z.txt  daemon=xrootd
INFO[2023-11-06T17:26:34Z] 231106 17:26:34 3218 ofs_open: unknown.2:36@fae8c2865de4 Unable to create /pelican/monitoring/director-test-2023-11-06T17:26:34Z.txt; permission denied  daemon=xrootd
INFO[2023-11-06T17:26:34Z] 231106 17:26:34 3218 unknown.2:36@fae8c2865de4 Xrootd_Response: sending err 3010: Unable to create /pelican/monitoring/director-test-2023-11-06T17:26:34Z.txt; permission denied  daemon=xrootd
INFO[2023-11-06T17:26:34Z] 231106 17:26:34 3218 XrootdXeq: unknown.2:36@fae8c2865de4 disc 0:00:00 (send failure)  daemon=xrootd
```

Regarding the errors from the origin, I looked up the reason `SSL peer certificate or SSH remote key was not OK` which seems to be related to the public key but I'm not sure how to resolve it. 

@bbockelm Any thoughts on how I should navigate?

**Comment by haoming29** [comment]
OK. After some investigation, I think this is an issue with self-signed CA and I shouldn't add director's public key to origin's `issuer.jwks` anyway because the intent should be let xrootd to query token's `iss` endpoint to get the issuer's JWK.

I believe in querying director's `iss` endpoint when scitoken tries to verify the token, there' s an SSL/SSH error and scitoken can't establish the connection to director's discovery endpoint. When I change the `iss` claim in director's token from `directorURL` to `originURL`, the error goes away and I get success for director's file upload to origin.

**Comment by bbockelm** [comment]
@haoming29 - I think you need to rebase onto `main`.  Looks like you're picking up commits from an older branch.

**Comment by bbockelm** [comment]
> OK. After some investigation, I think this is an issue with self-signed CA and I shouldn't add director's public key to origin's issuer.jwks anyway because the intent should be let xrootd to query token's iss endpoint to get the issuer's JWK.

I don't think it's an issue with the self-signed CA.  In that case, you usually get a different error message (more along the lines of "this CA isn't trusted").

> When I change the iss claim in director's token from directorURL to originURL, the error goes away and I get success for director's file upload to origin.

Ok, that's a hint.  What is the director URL versus the origin URL?  Is it possible the director URL is something like `https://localhost:8444` (not matching the certificate) whereas the origin URL is `https://fae8c2865de4:8443` (matching the certificate)?

File: director/origin_api.go
Comment (bbockelm) [reply director/origin_api.go:196]: My $0.02:
- There's no problem with using custom scopes; this is commonly done.  The WLCG profile just outlines the known ones, it doesn't ban unknown ones.
- This is an internal API so we have flexibility in the future.  However, if we change to use WLCG tokens, we will need to think about version compatibility -- will an old director still work with a new origin?

So, provided that we're OK with potentially needing compatibility code, I'm OK with leaving this as-is.

**Comment by haoming29** [comment]
> > OK. After some investigation, I think this is an issue with self-signed CA and I shouldn't add director's public key to origin's issuer.jwks anyway because the intent should be let xrootd to query token's iss endpoint to get the issuer's JWK.
> 
> I don't think it's an issue with the self-signed CA. In that case, you usually get a different error message (more along the lines of "this CA isn't trusted").
> 
> > When I change the iss claim in director's token from directorURL to originURL, the error goes away and I get success for director's file upload to origin.
> 
> Ok, that's a hint. What is the director URL versus the origin URL? Is it possible the director URL is something like `https://localhost:8444` (not matching the certificate) whereas the origin URL is `https://fae8c2865de4:8443` (matching the certificate)?

Yes! Hostname is the catch.

**Review by turetske — APPROVED** [review]
So this LGTM, but I want another pair of eyes to look at it as well.

File: origin_ui/origin_api.go
Comment (bbockelm) [inline origin_ui/origin_api.go:169]: As follow-up work, it's probably better to provide `ConfigureOriginAPI` a context object that can be cancelled on shutdown.

Having signal handlers scattered throughout the code is a bit scary -- too much indirect action -- and are fairly difficult to unit test compared to a context.

File: origin_ui/origin_api.go
Comment (bbockelm) [inline origin_ui/origin_api.go:141]: For future work -- I hate all the strings in this interface.  We should cast it to an enum.

IIRC, the reason why `SetComponentHealthStatus` returns an error is because it internally casts to an enum.  If we forced type safety on the caller, then we could eliminate the error return and simplify the code base.

File: origin_ui/origin_api.go
Comment (bbockelm) [inline origin_ui/origin_api.go:41]: Future work: this might be better-served as a Unix timestamp (and hence an unsigned 64-bit integer).

File: director/authentication.go
Comment (bbockelm) [inline director/authentication.go:17]:
```suggestion
const (
```

File: xrootd/authorization.go
Comment (bbockelm) [inline xrootd/authorization.go:308]: In the future, please avoid spurious whitespace changes.

File: director/origin_monitor.go
Comment (bbockelm) [inline director/origin_monitor.go:106]:
```suggestion
		return "", errors.Wrap(err, "Failed to create a token for the director test upload")
```

File: director/origin_monitor.go
Comment (bbockelm) [inline director/origin_monitor.go:289]: Future work: I don't like how much duplication there is between here and the origin self-test.  Feels like it's something that could have been factored out.

File: director/origin_monitor.go
Comment (bbockelm) [inline director/origin_monitor.go:286]: Better to remove this and change to `defer serverAdMutex.RUnlock()`.  If anything in the control block above panics, you would end up with the lock held forever and potentially deadlock in the future.

File: director/origin_monitor.go
Comment (bbockelm) [inline director/origin_monitor.go:281]: This means that, if an origin is decommissioned, we'd still get a log line about it forever, right?  Seems better to timeout...

File: director/redirect.go
Comment (bbockelm) [inline director/redirect.go:401]: Defer this instead.

File: director/redirect.go
Comment (bbockelm) [inline director/redirect.go:398]: What happens if the origin disappears from the cache, the cache gets garbage collected, then it comes up again?  As far as this logic goes, we now have a "new" origin and thus a second goroutine testing it.

Why not pass the go routine a `context.WithCancel(context.Background()), keep a cancel function in the cache, and then call the `cancel()` on invalidation?  That would remove the running goroutine when caches disappear.

**Review by bbockelm — CHANGES_REQUESTED** [review]
Almost there.

Review notes a number of "future work" items.  You're welcome to either fix them or create a follow-up issue with them noted.

The critical item is thinking about how to better stop testing origins when they disappear from the director... I think there's some bugs around the current code.

File: director/redirect.go
Comment (haoming29) [reply director/redirect.go:398]: I thought the logic in `origin_monitor.go` will return and end the goroutine when origin disappears from the cache.

File: director/origin_monitor.go
Comment (haoming29) [reply director/origin_monitor.go:281]: I didn't get why there would be a log line if the origin is evicted from the cache. But I agree it's safer and more robust to end the goroutine in `onEviction` handler of the cache. Will make the changes.

**Comment by haoming29** [comment]
With code review feedback, I will integrate most of the suggestions in this PR but leave the followings as future works and will create separate PRs for them:
* Refactor file test logic into a central place
* Provide `ConfigureOriginAPI` a context object that can be cancelled on shutdown
* Refactor `SetComponentHealthStatus` to use enum as status code instead of string literals

File: director/cache_ads_test.go
Comment (haoming29) [inline director/cache_ads_test.go:237]: As comment says, any thoughts on improving this is very welcomed.

File: director/redirect.go
Comment (haoming29) [reply director/redirect.go:401]: Introduced a new map to keep all cancel functions as the cache key needs to be comparable while the cancelFunc is not.

File: director/cache_ads_test.go
Comment (bbockelm) [reply director/cache_ads_test.go:237]: I would suggest something like this:

```
                 cancelChan := make(chan int)
		go func() {
			<-ctx.Done()
			if ctx.Err() == context.Canceled {
				cancelChan <- 1
			}
		}()

                 ...

                 select {
                 case <-cancelChan:
                 	require.True(t, true)
                 case <-time.After(3 * time.Second):
                 	require.True(t, false)
                 }
```

So, instead of polling (and in this case, once), you get a notification via the channel from the other go routine that the eviction is done.  It also includes a reasonable timeout so, on failure, the test doesn't hang forever.

**Review by bbockelm — APPROVED** [review]

**Pull request merged**
